File: src/main/java/club/asynclab/asyncraft/asyncauth/network/NetworkHandler.java
package club.asynclab.asyncraft.asyncauth.network;

import club.asynclab.asyncraft.asyncauth.AsyncAuth;
import club.asynclab.asyncraft.asyncauth.network.packet.ClientInitializePacket;
import club.asynclab.asyncraft.asyncauth.network.packet.ClientQuitPacket;
import club.asynclab.asyncraft.asyncauth.network.packet.LoginRequestPacket;
import club.asynclab.asyncraft.asyncauth.network.packet.LoginResponsePacket;
import net.minecraft.network.chat.TextComponent;
import net.minecraft.network.chat.TranslatableComponent;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.server.level.ServerPlayer;
import net.minecraftforge.network.NetworkDirection;
import net.minecraftforge.network.NetworkRegistry;
import net.minecraftforge.network.simple.SimpleChannel;

import java.util.Optional;

public class NetworkHandler {
    private static final String PROTOCOL = "1.0";
    public static SimpleChannel INSTANCE;
    private static int ID = 0;

    public static void init() {
        INSTANCE = NetworkRegistry.newSimpleChannel(
                new ResourceLocation(AsyncAuth.MODID, "main"),
                () -> PROTOCOL,
                PROTOCOL::equals,
                PROTOCOL::equals
        );
    }

    public static void registerPackets() {
        INSTANCE.registerMessage(ID++, ClientInitializePacket.class,
                ClientInitializePacket::encode,
                ClientInitializePacket::decode,
                ClientInitializePacket::handle
        );
        INSTANCE.registerMessage(ID++, LoginRequestPacket.class,
                LoginRequestPacket::encode,
                LoginRequestPacket::decode,
                LoginRequestPacket::handle
        );

        INSTANCE.registerMessage(ID++, LoginResponsePacket.class,
                LoginResponsePacket::encode,
                LoginResponsePacket::decode,
                LoginResponsePacket::handle
        );

        INSTANCE.registerMessage(ID++, ClientQuitPacket.class,
                (msg,buf) -> {},
                buf -> new ClientQuitPacket(),
                (msg,ctx) -> {
                    ServerPlayer player = ctx.get().getSender();
                    if (player != null)
                        player.connection.disconnect(new TextComponent("UnAuthorized"));
                }
        );
    }

}

File: forge-1.18.2/src/main/kotlin/club/asynclab/asyncraft/asyncauth/network/NetworkHandler.kt
package club.asynclab.asyncraft.asyncauth.network

import club.asynclab.asyncraft.asyncauth.AsyncAuth
import club.asynclab.asyncraft.asyncauth.misc.ModSetting
import club.asynclab.asyncraft.asyncauth.network.packet.auth.PacketLogin
import club.asynclab.asyncraft.asyncauth.network.packet.auth.PacketRegister
import club.asynclab.asyncraft.asyncauth.network.packet.auth.PacketResponse
import club.asynclab.asyncraft.asyncauth.network.packet.login.BasePacketLogin
import club.asynclab.asyncraft.asyncauth.network.packet.login.PacketFinish
import club.asynclab.asyncraft.asyncauth.network.packet.login.PacketInit
import club.asynclab.asyncraft.asyncauth.network.packet.misc.PacketHeart
import net.minecraftforge.fml.event.lifecycle.FMLCommonSetupEvent
import net.minecraftforge.network.HandshakeHandler
import net.minecraftforge.network.NetworkRegistry
import net.minecraftforge.network.simple.SimpleChannel


object NetworkHandler {
    private const val VERSION = "1.0"
    private var id = 0

    val LOGIN: SimpleChannel = NetworkRegistry.newSimpleChannel(
        AsyncAuth.resourceLocation("login"),
        { VERSION },
        VERSION::equals,
        VERSION::equals
    )

    fun onCommonSetup(event: FMLCommonSetupEvent) {
        LOGIN.messageBuilder(PacketInit::class.java, id++)
            .loginIndex(BasePacketLogin::loginIndex, BasePacketLogin::loginIndex::set)
            .encoder(PacketInit::encode)
            .decoder(PacketInit::decode)
            .consumer(PacketInit::handle)
            .buildLoginPacketList { local ->
                if (local)
                    emptyList()
                else
                    listOf(org.apache.commons.lang3.tuple.Pair.of("init", PacketInit(ModSetting.timeout.get())))
            }
            .add()
        LOGIN.messageBuilder(PacketLogin::class.java, id++)
            .encoder(PacketLogin::encode)
            .decoder(PacketLogin::decode)
            .consumer(PacketLogin::handle)
            .add()
        LOGIN.messageBuilder(PacketRegister::class.java, id++)
            .encoder(PacketRegister::encode)
            .decoder(PacketRegister::decode)
            .consumer(PacketRegister::handle)
            .add()
        LOGIN.messageBuilder(PacketResponse::class.java, id++)
            .encoder(PacketResponse::encode)
            .decoder(PacketResponse::decode)
            .consumer(PacketResponse::handle)
            .add()
        LOGIN.messageBuilder(PacketHeart::class.java, id++)
            .encoder(PacketHeart::encode)
            .decoder(PacketHeart::decode)
            .consumer(PacketHeart::handle)
            .add()
        LOGIN.messageBuilder(PacketFinish::class.java, id++)
            .loginIndex(BasePacketLogin::loginIndex, BasePacketLogin::loginIndex::set)
            .encoder(PacketFinish::encode)
            .decoder(PacketFinish::decode)
            .consumer(HandshakeHandler.indexFirst { _, packet, ctx ->
                PacketFinish.handle(packet, ctx)
            })
            .add()
    }
}